pkg/splunk/enterprise/upgrade.go

```diff
@@ -0,0 +1,315 @@
+		}
+
+		// check if the stateful set is created at this instance
+		statefulSet := &appsv1.StatefulSet{}
+		err := c.Get(ctx, namespacedName, statefulSet)
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				return true, nil
+			}
+			return false, nil
+		}
+		return true, nil
+	} else {
+
+		// check if a MonitoringConsole is attached to the instance
+		monitoringConsoleRef := spec.MonitoringConsoleRef
+		if monitoringConsoleRef.Name == "" {
+			goto SearchHeadCluster
+		}
+
+		namespacedName := types.NamespacedName{Namespace: cr.GetNamespace(), Name: monitoringConsoleRef.Name}
+		monitoringConsole := &enterpriseApi.MonitoringConsole{}
+
+		// get the monitoring console referred in custom resource
+		err := c.Get(ctx, namespacedName, monitoringConsole)
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				goto SearchHeadCluster
+			}
+			eventPublisher.Warning(ctx, "UpgradePathValidation", fmt.Sprintf("Could not find the Monitoring Console. Reason %v", err))
+			scopedLog.Error(err, "Unable to get Monitoring Console")
+			return false, err
+		}
+
+		mcImage, err := getCurrentImage(ctx, c, monitoringConsole, SplunkMonitoringConsole)
+		if err != nil {
+			eventPublisher.Warning(ctx, "UpgradePathValidation", fmt.Sprintf("Could not get the Monitoring Console Image. Reason %v", err))
+			scopedLog.Error(err, "Unable to get Monitoring Console current image")
+			return false, err
+		}
+
+		// check if an image upgrade is happening and whether MC has finished updating yet, return false to stop
+		// further reconcile operations on custom resource until MC is ready
+		if monitoringConsole.Status.Phase != enterpriseApi.PhaseReady || mcImage != spec.Image {
+			return false, nil
+		}
+
+		goto SearchHeadCluster
+	}
+SearchHeadCluster:
+	if cr.GroupVersionKind().Kind == "SearchHeadCluster" {
+
+		namespacedName := types.NamespacedName{
+			Namespace: cr.GetNamespace(),
+			Name:      GetSplunkStatefulsetName(SplunkSearchHead, cr.GetName()),
+		}
+
+		// check if the stateful set is created at this instance
+		statefulSet := &appsv1.StatefulSet{}
+		err := c.Get(ctx, namespacedName, statefulSet)
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				return true, nil
+			}
+			return false, nil
+		}
+		return true, nil
+	} else {
+
+		// get the clusterManagerRef attached to the instance
+		clusterManagerRef := spec.ClusterManagerRef
+
+		// check if a search head cluster exists with the same ClusterManager instance attached
+		searchHeadClusterInstance := enterpriseApi.SearchHeadCluster{}
+		opts := []rclient.ListOption{
+			rclient.InNamespace(cr.GetNamespace()),
+		}
+		searchHeadList, err := getSearchHeadClusterList(ctx, c, cr, opts)
+		if err != nil {
+			if err.Error() == "NotFound" {
+				goto IndexerCluster
+			}
+			return false, err
+		}
+		if len(searchHeadList.Items) == 0 {
+			goto IndexerCluster
+		}
+
+		// check if instance has the ClusterManagerRef defined
+		for _, shc := range searchHeadList.Items {
```

Comment:
Same comment as above, what if we have more than 1 SHC pointing to a CM? We might need to run through all the SHC in a list and make sure all of them are ready/upgraded.

Reply:
this function will run for every CR in Search Head Cluster. if the SHC is pointing to different CM, it should take in to consider that option.